src/keypair.rs

```diff
@@ -263,6 +274,45 @@ impl<G: Group> SizedBytes for PrivateKey<G> {
     }
 }
 
+/// A trait specifying the requirements for a private key container
+pub trait SecretKey<G: Group>: Clone + Sized + Zeroize {
+    /// Diffie-Hellman key exchange implementation
+    fn diffie_hellman(&self, pk: PublicKey<G>) -> Result<Vec<u8>, InternalPakeError>;
+
+    /// Returns public key from private key
+    fn public_key(&self) -> Result<PublicKey<G>, InternalPakeError>;
+
+    /// Serialization into bytes
+    fn serialize(&self) -> Vec<u8>;
+
+    /// Deserialization from bytes
+    fn deserialize(input: &[u8]) -> Result<Self, ProtocolError>;
+}
+
+impl<G: Group> SecretKey<G> for PrivateKey<G> {
+    fn diffie_hellman(&self, pk: PublicKey<G>) -> Result<Vec<u8>, InternalPakeError> {
+        let pk_data = GenericArray::<u8, G::ElemLen>::from_slice(&pk.0[..]);
+        let point = G::from_element_slice(pk_data)?;
+        let secret_data = GenericArray::<u8, G::ScalarLen>::from_slice(&self.0[..]);
+        Ok(G::mult_by_slice(&point, secret_data).to_arr().to_vec())
+    }
```

Comment:
I didn't realize this is all that DH is, is this correct? In any case, I'm about to test it soon.

Reply:
I confirmed this now.